### PR TITLE
Generator codelab

### LINKED ISSUE #273: Custom generator codelab
**Category**

Codelab

**Component**

Building a Custom Generator

**Is your feature request related to a problem? Please describe.**

We have no instructions on how to do this.

**Describe the solution you'd like**

Title: Building a custom generator

What you'll learn: 
- How to define a custom generator
- How to use your custom generator in a playground

**Describe alternatives you've considered**

More documentation on devsite.

**Additional context**

See [Writing a Codelab](https://github.com/google/blockly-samples/wiki/Writing-a-Codelab)

Next steps: 
- Flesh out the outline
- Decide what the custom generator will do and what language it will generate?

### PR DESCRIPTION
**Pull request by rachel-fenichel**
Fixes #273 

Use "view rich diff" to review this.

This first draft has:
- introduction
- problem statement
- block definitions
- description of language generators
- generators for some blocks

Not written yet:
- array block generator
- object block generator
- putting it all together
- summary
- resources

### CODE REVIEW

File: codelabs/custom_generator/custom_generator.md [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+```js
+myGenerator['logic_boolean'] = function(block) {
+  const code = (block.getFieldValue('BOOL') == 'TRUE') ? 'true' : 'false';
+  return [code, myGenerator.PRECEDENCE];
+```
+
+### Summary
+
+- Value blocks return an array containing the value as a string and the precedence.
+- `getFieldValue` find the field with the specified name and returns its value.
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md]: finds

File: codelabs/custom_generator/custom_generator.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+<block type="member"></block>
+<block type="math_number"><field name="NUM">0</field></block>
+<block type="text"><field name="TEXT"/></block>
+<block type="logic_boolean"><field name="BOOL">TRUE</field></block>
+<block type="logic_null"/>
+<block type="lists_create_with"><mutation items="3"/></block></xml>
+`
+```
+
+Then update the options struct in the playgound to use your new toolbox:
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md]: playground

File: codelabs/custom_generator/custom_generator.md [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+Feedback Link: https://github.com/google/blockly-samples/issues/new
+
+# Build a custom generator
+
+## Codelab overview
+
+### What you'll learn
+- How to create a custom language generator.
+- How to create block generator definitions for existing blocks.
+- How to create block generator definitions for new block.
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md]: a new block, or new blocks.

File: codelabs/custom_generator/custom_generator.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+```js
+var codelabToolbox = `
+<xml id="toolbox" style="display: none">
+<block type="object"/>
+<block type="member"></block>
+<block type="math_number"><field name="NUM">0</field></block>
+<block type="text"><field name="TEXT"/></block>
+<block type="logic_boolean"><field name="BOOL">TRUE</field></block>
+<block type="logic_null"/>
+<block type="lists_create_with"><mutation items="3"/></block></xml>
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md]: Add a new line before </xml> for clarity.

File: codelabs/custom_generator/custom_generator.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+`
+```
+
+Then update the options struct in the playgound to use your new toolbox:
+
+```js
+var defaultOptions = {
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md:130]: Given you're using the playground.html for this codelab, I wonder if: 
- we update our playground.html to have a defaultOptions struct rather than as it is right now (inline)
- update the codelab to use the advanced playground
- (just a thought) maybe we build / use an interactive environment similar to https://observablehq.com/@samelhusseini/blockly-notebook

File: codelabs/custom_generator/custom_generator.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+console.log(generatedCode);
+```
+
+Note: You must use `workspaceToCode` as the entry point for code generation, passing in the workspace. `workspaceToCode` calls `init` and `finish` to do any additional setup and teardown that your generator needs.
+
+Other functions such as `blockToCode` are public so that you can use them in your per-block generator definitions, but must not be used as the entry point.
+
+### Test it
+
+Put a number block on the workspace, then click your button to call your generator.
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md:184]: A screenshot here to help follow along

Comment (samelhusseini) [reply codelabs/custom_generator/custom_generator.md:184]: Also, what button, and what's it calling? Perhaps refer to the button in the previous step as the "Generate" button, and then something like: "then click the Generate button to call your generator's workspaceToCode method"

File: codelabs/custom_generator/custom_generator.md [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+Check the console for the output. You should see an error:
+
+```
+generator.js:182 Uncaught Error: Language "Codelab" does not know how to generate  code for block type "math_number".
+    at Blockly.Generator.blockToCode (generator.js:182)
+    at Blockly.Generator.workspaceToCode (generator.js:94)
+    at <anonymous>:1:18
+```
+
+This error occurs because you need to write a generator for each type of block. Read the next section for more details.
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md]: "generator" in that sentence is a little overloaded as we refer to the whole thing as a generator. Perhaps: "you need to write a code generator for each type of block"

Comment (samelhusseini) [reply codelabs/custom_generator/custom_generator.md]: https://developers.google.com/blockly/guides/create-custom-blocks/generating-code refers to them as code generators too

File: codelabs/custom_generator/custom_generator.md [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+You need to be able to access this value inside your block generators, so add `PRECEDENCE` to your generator:
+
+```js
+myGenerator.PRECEDENCE = 0;
+```
+
+## Value block generators
+
+In this step you will build the generators for the simple value blocks: null, string, number, and boolean.
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md]: surround null, string, number, boolean with backticks.

File: codelabs/custom_generator/custom_generator.md [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+```js
+myGenerator['member'] = function(block) {
+  const text_member_name = block.getFieldValue('MEMBER_NAME');
+  const value_name = codelabGenerator.valueToCode(block, 'MEMBER_VALUE',
+      codelabGenerator.PRECEDENCE);
+  const code = '"' + text_member_name + '" : ' + value_name + ',\n';
+  return code;
+};
+```
+
```
Comment (samelhusseini) [inline codelabs/custom_generator/custom_generator.md:390]: Will you add one for myGenerator['object'] ?

**Review by samelhusseini — COMMENTED** [review]
Looking good, added comments.

File: codelabs/custom_generator/custom_generator.md (hunk 5 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md:130]: That's one of my big questions: where do I tell them to set this up?
I'm leaning toward using the advanced playground so I can get the benefits of our dev tools. The only downside I see is that I might have to update this codelab shortly after publishing it, especially if your playground work gets to the point where it basically replaces the advanced playground.

This same issue shows up where I say to add a button to generate code: if your playground includes setups for adding new generators, do I need to have code for them to add a button? So far I've just been logging to the console, but really I should use the text area instead.

re: observablehq: ooh, that's nifty

File: codelabs/custom_generator/custom_generator.md (hunk 6 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md:184]: Screenshot: definitely. I'm adding a todo for now, because it'll depend on which playground I'm using.

File: codelabs/custom_generator/custom_generator.md (hunk 7 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md]: I've been using "generator" to refer to the instance of Blockly.Generator, and "block generator" for a block's code generation function. Let's discuss here and pick preferred wording, and I'll update through the whole codelab.

Some options:
- "language generator" for the instance of Blockly.Generator, and "block generator" for the individual functions
- "generator", "block generator"
- "language generator", "code generator"
- "generator", "code generator"
- "language generator", "block code generator"
- "generator", "block code generator"

File: codelabs/custom_generator/custom_generator.md (hunk 5 above)
Comment (samelhusseini) [reply codelabs/custom_generator/custom_generator.md:130]: Agreed that we need a consistent strategy for where one follows along. The advanced playground, aside from its name (might be unfriendly to beginners), makes sense to me as we can setup things like mechanisms for registering a new generator, adding a button, etc.

File: codelabs/custom_generator/custom_generator.md (hunk 7 above)
Comment (samelhusseini) [reply codelabs/custom_generator/custom_generator.md]: "language generator", "code generator" is how I think our documentation refers to these two. Don't really have a strong opinion on this, but maybe: 
"language generator", and "block-to-code generator"?

**Comment by rachel-fenichel** [comment]
This PR is out of draft and ready for review.

File: codelabs/custom_generator/custom_generator.md [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+This codelab assumes that you are already comfortable with using the Blockly playground locally.
+
+## Setup
+
+In this codelab you will add code to the Blockly playground to create and use a new generator.
+
+### The playground
+
+You will make all of your changes in the advanced playground, which you can find at `tests/advanced_playground.html`. This playground contains all of Blockly's base blocks, as well as some developer tools to make testing easier.
```
Comment (alschmiedt) [inline codelabs/custom_generator/custom_generator.md]: Right now the advanced_playground is in tests/playgrounds/advanced_playground.html

File: codelabs/custom_generator/custom_generator.md [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+Block generators are defined on the language generator object. For instance, here is the code to add a block generator for blocks of type `sample_block` on a language generator object (`sampleGenerator`).
+
+```js
+sampleGenerator['sample_block'] = function(block) {
+  return 'my code string';
+}
+```
+
+### Statement blocks
+Statement blocks represent code that does not have an output connection.
```
Comment (alschmiedt) [inline codelabs/custom_generator/custom_generator.md]: Add space after header for consistency & might be nice to have an image. Some people might not know what an output connection is.

File: codelabs/custom_generator/custom_generator.md [hunk 12]
```diff
@@ -0,0 +1,602 @@
...
+The `logic_boolean` block has a dropdown field named `BOOL`.
+
+![](./boolean_block.png)
+
+Calling `getFieldValue` on a dropdown field returns the value of the selected option, which may not be the same as the display text. In this case the dropdown has two possible values: `TRUE` and `FALSE`.
+
+```js
+codelabGenerator['logic_boolean'] = function(block) {
+  const code = (block.getFieldValue('BOOL') == 'TRUE') ? 'true' : 'false';
+  return [code, codelabGenerator.PRECEDENCE];
```
Comment (alschmiedt) [inline codelabs/custom_generator/custom_generator.md:322]: Add a };

File: codelabs/custom_generator/custom_generator.md [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+```
+
+As with member blocks, there are no restrictions on the types of blocks connected to inputs.
+
+### Gather values
+
+Each value input on the block has a name: `ADD0`, `ADD1`, etc. Use `valueToCode` in a loop to build an array of values:
+
+```js
+const elements = [];
```
Comment (alschmiedt) [inline codelabs/custom_generator/custom_generator.md]: elements -> values to match rest of section

File: codelabs/custom_generator/custom_generator.md [hunk 14]
```diff
@@ -0,0 +1,602 @@
...
+  }
+  const valueString = values.join(',\n');
+  const indentedValueString =
+      codelabGenerator.prefixLines(valueString, codelabGenerator.INDENT);
+  const codeString = '[\n' + indentedValueString + '\n]';
+  return [codeString, codelabGenerator.PRECEDENCE];
+};
+```
+
+Test the block generator by adding an array to your onscreen blocks an populating it.
```
Comment (alschmiedt) [inline codelabs/custom_generator/custom_generator.md]: an -> and

Comment (alschmiedt) [reply codelabs/custom_generator/custom_generator.md]: add "Test It" header

File: codelabs/custom_generator/custom_generator.md [hunk 15]
```diff
@@ -0,0 +1,602 @@
...
+
+Create a stack of `member` blocks on the workspace and click "To JSON". You should see generated code for all of your blocks, not just the first one.
+
+Next, add an `object` block and drag your `member` blocks into it, then click "To JSON". This case tests `statementToCode`, and should still generated code for all of your blocks.
+
+## Summary
+
+In this codelab you:
+- Built a custom language generator to generate JSON.
+- Defined block generators for built in blocks and for custom blocks
```
Comment (alschmiedt) [inline codelabs/custom_generator/custom_generator.md]: nit: add .

File: codelabs/custom_generator/custom_generator.md [hunk 16]
```diff
@@ -0,0 +1,602 @@
...
+- Learned how to use the core generator functions: `statementToCode`, `valueToCode`, `blockToCode`, and `getFieldValue`.
+- Learned how to generate code for stacks of blocks.
+
+JSON is a simple language, and there are many additional features you may want to implement in your generator. Blockly's built-in language generators are a good place to learn more about some additional features:
+- Variable definition and use.
+- Function definition and use.
+- Initialization and cleanup.
+- Injecting additional functions and variables.
+
+Blockly ships with five language generators: Python, Dart, JavaScript, PHP, and Lua. You can find the language generators and block generators in the `generators` directory.
```
Comment (alschmiedt) [inline codelabs/custom_generator/custom_generator.md]: Add a link?

**Review by alschmiedt — COMMENTED** [review]
This looks great!

File: codelabs/custom_generator/custom_generator.md (hunk 10 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md]: Done

File: codelabs/custom_generator/custom_generator.md (hunk 6 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md:184]: I ended up with no screenshot here because it didn't feel helpful, but instead I added much more explicit instructions on how and where to add the button and its callback.

File: codelabs/custom_generator/custom_generator.md (hunk 11 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md]: Actually, I'm going to rephrase this to "represent code that does not return a value".

That way it focuses more on the code that it represents.

I'm assuming they know how to build custom blocks, so I think they'll know what an output connection is.

File: codelabs/custom_generator/custom_generator.md (hunk 12 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md:322]: Done.

File: codelabs/custom_generator/custom_generator.md (hunk 9 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md:390]: Done.

File: codelabs/custom_generator/custom_generator.md (hunk 13 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md]: Done.

File: codelabs/custom_generator/custom_generator.md (hunk 14 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md]: Done and done.

File: codelabs/custom_generator/custom_generator.md (hunk 15 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md]: Done.

File: codelabs/custom_generator/custom_generator.md (hunk 16 above)
Comment (rachel-fenichel) [reply codelabs/custom_generator/custom_generator.md]: Done.

**Review by alschmiedt — APPROVED** [review]
